app/models/relation/hierarchy_paths.rb

```diff
@@ -86,21 +86,49 @@ def alter_hierarchy_path
     end
 
     def self.add_hierarchy_path_sql(id = nil)
-      <<-SQL
-        INSERT INTO
-          #{hierarchy_table_name}
-          (work_package_id, path)
-        SELECT
-          to_id, #{add_hierarchy_agg_function} AS path
-          FROM
-          (SELECT to_id, from_id, hierarchy
-          FROM relations
-          WHERE hierarchy > 0 AND relates = 0 AND blocks = 0 AND duplicates = 0 AND includes = 0 AND requires = 0 AND follows = 0
-          #{add_hierarchy_id_constraint(id)}
-          ) ordered_by_hierarchy
+      # Added a slighly optimized sql statement, which works without subselects. STEFFEN
+      if id.nil?
+        stmt = <<-SQL
+          INSERT INTO
+            #{hierarchy_table_name}
+            (work_package_id, path)
+          SELECT
+            to_id, #{add_hierarchy_agg_function} AS path
+            FROM
+            (SELECT to_id, from_id, hierarchy
+               FROM relations
+               WHERE hierarchy > 0 AND relates = 0 AND blocks = 0 AND duplicates = 0 AND includes = 0 AND requires = 0 AND follows = 0
+            ) ordered_by_hierarchy
+            GROUP BY to_id
+            #{add_hierarchy_conflict_statement}
+        SQL
+      else
+        stmt = <<-SQL
+          INSERT INTO
+            #{hierarchy_table_name}
+            (work_package_id, path)
+          SELECT
+            to_id, #{add_hierarchy_agg_function} AS path
+          FROM (
+            SELECT to_id, from_id, hierarchy
```

Comment:
As the statements between having an id and not having only differs within the subselect, it makes sense to extract only those differences into a separate method.